### PR TITLE
[hotwired__turbo] Make `FrameElement.src` nullable

### PR DESCRIPTION
**Pull request by marcoroth**
## What

`FrameElement.src` is currently typed as `string`, but it is nullable on both read and write.

## Why

Looking at Turbo's [`frame_element.js`](https://github.com/hotwired/turbo/blob/v8.0.23/src/elements/frame_element.js#L58-L74):

```js
get src() {
  return this.getAttribute("src")
}

set src(value) {
  if (value) {
    this.setAttribute("src", value)
  } else {
    this.removeAttribute("src")
  }
}
```

- The **getter** returns `getAttribute("src")`, which is `null` for a frame without a `src` attribute — a normal state for eager/lazy frames before a source is set.
- The **setter** accepts `null` (or any falsy value) and removes the attribute, which is the idiomatic way to clear a frame's source.

Turbo's own bundled [`.d.ts` (before they were dropped in Turbo 8) declared this as `string | null`](https://npmx.dev/package-code/@hotwired/turbo/v/7.3.0/dist%2Ftypes%2Felements%2Fframe_element.d.ts#L34-L35). The current definition narrows it to `string`, which both forces spurious non-null assumptions when reading `.src` and rejects the valid `frame.src = null` / `frame.src = ""` clearing pattern.

## Changes

- Widen `FrameElement.src` from `string` to `string | null`.
- Add tests asserting the `string | null` type and that both `string` and `null` are assignable to the setter.

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - https://github.com/hotwired/turbo/blob/v8.0.23/src/elements/frame_element.js#L58-L74
  - https://npmx.dev/package-code/@hotwired/turbo/v/7.3.0/dist%2Ftypes%2Felements%2Fframe_element.d.ts#L34-L35
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/hotwired__turbo/index.d.ts
Comment (marcoroth) [inline types/hotwired__turbo/index.d.ts:6]: Actually, thinking about it more, I guess all custom attributes are technically nullable.

Comment (myabc) [reply types/hotwired__turbo/index.d.ts:6]: I looked at the implementation. `refresh` is currently missing (also nullable), while `isPreview` could return `boolean | undefined` (ideally this would be coerced to a boolean in the implementation for the sake of consistency!)

For `loading`, it looks like there is getter/setter asymmetry. `"eager" | "lazy"` vs `"eager" | "lazy" | null | undefined`. I'm more inclined _not_ to widen the type (and potentially confuse consumers), but I might be wrong.

```ts
// explicit
frameElement.loading = "eager"
// seems to make more sense than nullifying it to get the default
frameElement.loading = null
frameElement.loading = "" // who does this?
```

Perhaps @G-Rath has some thoughts?

File: types/hotwired__turbo/index.d.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 export class FrameElement extends HTMLElement {
-    src: string;
+    src: string | null;
     disabled: boolean;
```
Comment (myabc) [inline types/hotwired__turbo/index.d.ts:3]:
```suggestion
    refresh: string | null;
    disabled: boolean;
```

File: types/hotwired__turbo/hotwired__turbo-tests.ts [hunk 2]
```diff
@@ -42,6 +42,11 @@ turboFrame.loading = "slow";
 
 turboFrame.reload().catch(console.error);
 
+// $ExpectType string | null
+turboFrame.src;
+turboFrame.src = "/messages";
+turboFrame.src = null;
+
```
Comment (myabc) [inline types/hotwired__turbo/hotwired__turbo-tests.ts:49]:
```suggestion

// $ExpectType string | null
turboFrame.refresh;
turboFrame.refresh = "morph";
turboFrame.refresh = null;
```

**Review by myabc — CHANGES_REQUESTED** [review]
Thanks @marcoroth ! 

Looks good but if it's ok with you, let's take care of the missing `FrameElement#refresh` property.

**Review by G-Rath — APPROVED** [review]
this is definitely correct 🤷 

Image

**Comment by G-Rath** [comment]
fwiw I think it would be fine to land without adding `refresh` - that's easily done in another PR

File: types/hotwired__turbo/index.d.ts
Comment (G-Rath) [reply types/hotwired__turbo/index.d.ts:6]: > I guess all custom attributes are technically nullable.

if you mean in the sense that they can be assigned `null` without error, that's true of tons of stuff in JS and why you shouldn't type everything after what the runtime allows but what you intend to support 😝 

I've not looked at the implementation but hopefully there isn't an actual `return undefined`? 

`isPreview` and `loading` cannot differ from what is currently typed by assignment at least

Image

Comment (myabc) [reply types/hotwired__turbo/index.d.ts:6]: `isPreview` can return `undefined`, yes.

Comment (G-Rath) [reply types/hotwired__turbo/index.d.ts:6]: can you link to the code that does that? I'm curious to why

**Comment by myabc** [comment]
> fwiw I think it would be fine to land without adding refresh - that's easily done in another PR

I'm ok with that too.. I wasn't sure whether DefinitelyTyped did one release per merge, or given the slight delay in automated releases, it batches them.

**Review by myabc — APPROVED** [review]

File: types/hotwired__turbo/index.d.ts
Comment (myabc) [reply types/hotwired__turbo/index.d.ts:6]: it's down to the optional chaining:

```js
  get isPreview() {
    return this.ownerDocument?.documentElement?.hasAttribute("data-turbo-preview")
  }
```

a `?? false` should would take care of it ensuring it's always a boolean.

https://github.com/hotwired/turbo/blob/main/src/elements/frame_element.js#L182

**Comment by myabc** [comment]
Ready to merge

File: types/hotwired__turbo/index.d.ts
Comment (G-Rath) [reply types/hotwired__turbo/index.d.ts:6]: right - fwiw I don't think this can actually be `undefined` because `ownerDocument` is only ever `null` if [the node itself is `Document`](https://developer.mozilla.org/en-US/docs/Web/API/Node/ownerDocument#value) which it can't be because it should always be a `turbo-frame`

I would say for this we assume it cannot be null until someone actually reports an instance of that being the case 🤷

Comment (myabc) [reply types/hotwired__turbo/index.d.ts:6]: Sounds good. I dropped `isPreview` from follow-up PR #75196

**Comment by marcoroth** [comment]
Thanks @myabc and @G-Rath! 🙏🏼
